packages/light-client/src/block.ts

```diff
@@ -0,0 +1,165 @@
+        );
+    }
+
+    // (4) and (5)
+    // (4) `approvals_after_next` contains valid signatures on the block producer approval messages.
+    // (5) The signatures present represent more than 2/3 of the total stake.
+    const totalStake = new BN(0);
+    const approvedStake = new BN(0);
+
+    const currentBlockHash = computeBlockHash(newBlock);
+    const nextBlockHash = combineHash(
+        bs58.decode(newBlock.next_block_inner_hash),
+        currentBlockHash
+    );
+
+    for (let i = 0; i < blockProducers.length; i++) {
+        const approval = newBlock.approvals_after_next[i];
+        const stake = blockProducers[i].stake;
+
+        totalStake.iadd(new BN(stake));
+
+        if (approval === null) {
+            continue;
+        }
+
+        approvedStake.iadd(new BN(stake));
+
+        const publicKey = PublicKey.fromString(blockProducers[i].public_key);
+        const signature = bs58.decode(approval.split(':')[1]);
+
+        const approvalEndorsement = serialize(
+            SCHEMA,
+            new BorshApprovalInner({ endorsement: nextBlockHash })
+        );
+
+        const approvalHeight: BN = new BN(newBlock.inner_lite.height + 2);
+        const approvalHeightLe = approvalHeight.toArrayLike(Buffer, 'le', 8);
+        const approvalMessage = new Uint8Array([
+            ...approvalEndorsement,
+            ...approvalHeightLe,
+        ]);
+
+        if (!publicKey.verify(approvalMessage, signature)) {
+            throw new Error(
+                `Invalid approval message signature for validator ${blockProducers[i].account_id}`
+            );
+        }
+    }
+
+    // (5) Calculates the 2/3 threshold and checks that the approved stake accumulated above
+    // exceeds it.
+    const threshold = totalStake.mul(new BN(2)).div(new BN(3));
+    if (approvedStake.lte(threshold)) {
+        throw new Error('Approved stake does not exceed the 2/3 threshold');
+    }
+
+    // (6) Verify that if the new block is in the next epoch, the hash of the next block producers
+    // equals the `next_bp_hash` provided in that block.
+    if (
+        newBlock.inner_lite.epoch_id === lastKnownBlock.inner_lite.next_epoch_id
+    ) {
+        // (3) If the block is in a new epoch, then `next_bps` must be present.
+        if (!newBlock.next_bps) {
+            throw new Error(
+                'New block must include next block producers if a new epoch starts'
+            );
+        }
+
+        const bpsHash = hashBlockProducers(newBlock.next_bps);
+
+        if (!bpsHash.equals(bs58.decode(newBlock.inner_lite.next_bp_hash))) {
+            throw new Error('Next block producers hash doesn\'t match');
+        }
+    }
+}
+
+function hashBlockProducers(bps: ValidatorStakeView[]): Buffer {
+    const borshBps: BorshValidatorStakeView[] = bps.map((bp) => {
+        if (bp.validator_stake_struct_version) {
+            const version = parseInt(
+                bp.validator_stake_struct_version.slice(1)
```

Comment:
how about the more strict (and safer) check: `bp.validator_stake_struct_version === "V1"`